meld/cluster_topdown.py

```diff
@@ -0,0 +1,305 @@
+        """
+            Perform an incremental update to the summed TFCE integral over
+            cluster members
+
+            Args
+            ----
+            lower_limit : float
+                lower limit of TFCE integral
+                called "bottomVal" in Cpp implementation
+        """
+        if len(self.members) == 0:
+            self.upper_limit = lower_limit
+        elif not lower_limit == self.upper_limit:
+            assert lower_limit < self.upper_limit
+            new_slice_val = tfce_calc(self.extent,
+                                      self.param_e,
+                                      self.upper_limit,
+                                      self.h_plus1,
+                                      lower_limit)
+            self.accum_val += new_slice_val
+            self.upper_limit = lower_limit
+
+    def add_member(self, node, height, extent):
+        """
+            Add a new node to cluster
+
+            Args
+            ----
+            node : int
+                Node index
+            height : float
+                Test statistic value at node
+                called "val" in Cpp implementation
+            extent : float
+                Extent of node
+                called "area" in Cpp implementation
+        """
+        self.update(height)
+        # append node to list of members
+        self.members.append(node)
+        self.extent += extent
+
+
+def alloc_cluster(cluster_list, dead_clusters, param_e, param_h):
+    """
+        Generate a valid (new) cluster number
+        If available, number for a dead cluster will be recycled
+        Cluster and dead cluster lists will be updated
+
+        Args
+        ----
+            cluster_list : list of Clusters
+            dead_clusters : set of Clusters
+            param_e : float
+            param_h : float
+    """
+    # verify that the lists are actually being modified in main code
+    if len(dead_clusters) == 0:
+        cluster_list.append(Cluster(param_e, param_h))
+        return len(cluster_list) - 1
+    else:
+        # removes and returns element from set
+        ret = dead_clusters.pop()
+        cluster_list[ret] = Cluster(param_e, param_h)
+        return ret
+
+
+def get_node_neighbors(node, connectivity):
+    """
+        Given a connectivity matrix, get neighbors of a node
+
+        Args
+        ----
+            node : int
+                Node index
+            connectivity : sparse matrix in COO format
+
+        Returns
+        -------
+            neighbors : numpy array of integers
+                array of node indices for neighboring nodes
+
+    """
+    # since connectivity matrix is triangular, we need to check both along rows
+    # and cols
+    # diagonals are all 0 in connectivity matrix (no self-neighboring) so we
+    # don't need to worry about that
+    neighbors = connectivity[node]
+
+    return neighbors
+
+
+@jit(nopython=True)
+def get_touching(neighbors, membership):
```

Comment:
This function is also fairly simple, and I believe it is only called once, you might try removing it and putting the code inline, to see if it runs faster.

Reply:
again, good catch, I'll try that.